File: ad-revenue-ironsource-v7/src/test/java/io/appmetrica/analytics/adrevenue/ironsource/v7/impl/AdRevenueConverterTest.java
package io.appmetrica.analytics.adrevenue.ironsource.v7.impl;

import com.ironsource.mediationsdk.impressionData.ImpressionData;
import io.appmetrica.analytics.modulesapi.internal.client.adrevenue.AdRevenueConstants;
import io.appmetrica.analytics.modulesapi.internal.client.adrevenue.ModuleAdRevenue;
import io.appmetrica.analytics.modulesapi.internal.client.adrevenue.ModuleAdType;
import io.appmetrica.analytics.testutils.CommonTest;
import java.math.BigDecimal;
import java.util.Currency;
import java.util.HashMap;
import org.junit.Before;
import org.junit.Test;

import static io.appmetrica.analytics.assertions.AssertionsKt.ObjectPropertyAssertions;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class AdRevenueConverterTest extends CommonTest {

    private final double revenue = 123.1234;
    private final String adUnit = "Interstitial";
    private final String adNetwork = "adNetwork";
    private final String placement = "placement";
    private final String precision = "precision";

    private final ImpressionData data = mock(ImpressionData.class);

    private final AdRevenueConverter converter = new AdRevenueConverter();

    @Before
    public void setUp() {
        when(data.getRevenue()).thenReturn(revenue);
        when(data.getAdUnit()).thenReturn(adUnit);
        when(data.getAdNetwork()).thenReturn(adNetwork);
        when(data.getPlacement()).thenReturn(placement);
        when(data.getPrecision()).thenReturn(precision);
    }

    @Test
    public void convert() {
        final ModuleAdRevenue adRevenue = converter.convert(data);

        ObjectPropertyAssertions(adRevenue)
            .checkField("adRevenue", BigDecimal.valueOf(revenue))
            .checkField("currency", Currency.getInstance("USD"))
            .checkField("adType", ModuleAdType.INTERSTITIAL)
            .checkField("adNetwork", adNetwork)
            .checkFieldIsNull("adUnitId")
            .checkFieldIsNull("adUnitName")
            .checkFieldIsNull("adPlacementId")
            .checkField("adPlacementName", placement)
            .checkField("precision", precision)
            .checkField("payload", new HashMap<String, String>() {{
                put(AdRevenueConstants.ORIGINAL_SOURCE_KEY, "ad-revenue-ironsource-v7");
                put(AdRevenueConstants.ORIGINAL_AD_TYPE_KEY, "Interstitial");
                put(AdRevenueConstants.SOURCE_KEY, "ironsource");
            }})
            .checkField("autoCollected", true)
            .checkAll();
    }

    @Test
    public void convertIfRevenueIsIllegal() {
        when(data.getRevenue()).thenReturn(Double.NaN);
        final ModuleAdRevenue adRevenue = converter.convert(data);

        ObjectPropertyAssertions(adRevenue)
            .checkField("adRevenue", BigDecimal.valueOf(0.0))
            .checkField("currency", Currency.getInstance("USD"))
            .checkField("adType", ModuleAdType.INTERSTITIAL)
            .checkField("adNetwork", adNetwork)
            .checkFieldIsNull("adUnitId")
            .checkFieldIsNull("adUnitName")
            .checkFieldIsNull("adPlacementId")
            .checkField("adPlacementName", placement)
            .checkField("precision", precision)
            .checkField("payload", new HashMap<String, String>() {{
                put(AdRevenueConstants.ORIGINAL_SOURCE_KEY, "ad-revenue-ironsource-v7");
                put(AdRevenueConstants.ORIGINAL_AD_TYPE_KEY, "Interstitial");
                put(AdRevenueConstants.SOURCE_KEY, "ironsource");
            }})
            .checkField("autoCollected", true)
            .checkAll();
    }

    @Test
    public void convertAndCheckAutoAdType() {
        when(data.getAdUnit()).thenReturn(null);
        final ModuleAdRevenue adRevenueNull = converter.convert(data);
        assertThat(adRevenueNull.getAdType()).isNull();
        assertThat(adRevenueNull.getPayload()).isEqualTo(
            new HashMap<String, String>() {{
                put(AdRevenueConstants.ORIGINAL_SOURCE_KEY, "ad-revenue-ironsource-v7");
                put(AdRevenueConstants.ORIGINAL_AD_TYPE_KEY, "null");
                put(AdRevenueConstants.SOURCE_KEY, "ironsource");
            }}
        );

        when(data.getAdUnit()).thenReturn("Rewarded Video");
        final ModuleAdRevenue adRevenueRewarded = converter.convert(data);
        assertThat(adRevenueRewarded.getAdType()).isEqualTo(ModuleAdType.REWARDED);
        assertThat(adRevenueRewarded.getPayload()).isEqualTo(
            new HashMap<String, String>() {{
                put(AdRevenueConstants.ORIGINAL_SOURCE_KEY, "ad-revenue-ironsource-v7");
                put(AdRevenueConstants.ORIGINAL_AD_TYPE_KEY, "Rewarded Video");
                put(AdRevenueConstants.SOURCE_KEY, "ironsource");
            }}
        );

        when(data.getAdUnit()).thenReturn("Interstitial");
        final ModuleAdRevenue adRevenueInterstitial = converter.convert(data);
        assertThat(adRevenueInterstitial.getAdType()).isEqualTo(ModuleAdType.INTERSTITIAL);
        assertThat(adRevenueInterstitial.getPayload()).isEqualTo(
            new HashMap<String, String>() {{
                put(AdRevenueConstants.ORIGINAL_SOURCE_KEY, "ad-revenue-ironsource-v7");
                put(AdRevenueConstants.ORIGINAL_AD_TYPE_KEY, "Interstitial");
                put(AdRevenueConstants.SOURCE_KEY, "ironsource");
            }}
        );

        when(data.getAdUnit()).thenReturn("Banner");
        final ModuleAdRevenue adRevenueBanner = converter.convert(data);
        assertThat(adRevenueBanner.getAdType()).isEqualTo(ModuleAdType.BANNER);
        assertThat(adRevenueBanner.getPayload()).isEqualTo(
            new HashMap<String, String>() {{
                put(AdRevenueConstants.ORIGINAL_SOURCE_KEY, "ad-revenue-ironsource-v7");
                put(AdRevenueConstants.ORIGINAL_AD_TYPE_KEY, "Banner");
                put(AdRevenueConstants.SOURCE_KEY, "ironsource");
            }}
        );

        when(data.getAdUnit()).thenReturn("Some string");
        final ModuleAdRevenue adRevenueOther = converter.convert(data);
        assertThat(adRevenueOther.getAdType()).isEqualTo(ModuleAdType.OTHER);
        assertThat(adRevenueOther.getPayload()).isEqualTo(
            new HashMap<String, String>() {{
                put(AdRevenueConstants.ORIGINAL_SOURCE_KEY, "ad-revenue-ironsource-v7");
                put(AdRevenueConstants.ORIGINAL_AD_TYPE_KEY, "Some string");
                put(AdRevenueConstants.SOURCE_KEY, "ironsource");
            }}
        );
    }
}

File: ad-revenue-applovin-v12/src/test/java/io/appmetrica/analytics/adrevenue/applovin/v12/impl/AdRevenueConverterTest.kt
package io.appmetrica.analytics.adrevenue.applovin.v12.impl

import com.applovin.mediation.MaxAd
import com.applovin.mediation.MaxAdFormat
import com.applovin.sdk.AppLovinSdk
import com.applovin.sdk.AppLovinSdkConfiguration
import io.appmetrica.analytics.assertions.ObjectPropertyAssertions
import io.appmetrica.analytics.modulesapi.internal.client.adrevenue.AdRevenueConstants
import io.appmetrica.analytics.modulesapi.internal.client.adrevenue.ModuleAdRevenue
import io.appmetrica.analytics.modulesapi.internal.client.adrevenue.ModuleAdType
import io.appmetrica.analytics.testutils.CommonTest
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test
import org.mockito.kotlin.doReturn
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever
import java.math.BigDecimal
import java.util.Currency

internal class AdRevenueConverterTest : CommonTest() {

    private val revenue = 123.1234
    private val format = MaxAdFormat.REWARDED
    private val networkName = "some_networkName"
    private val placement = "some_placement"
    private val networkPlacement = "some_networkPlacement"
    private val revenuePrecision = "some_revenuePrecision"
    private val countryCode = "RU"

    private val maxAd: MaxAd = mock {
        on { revenue } doReturn revenue
        on { format } doReturn format
        on { networkName } doReturn networkName
        on { placement } doReturn placement
        on { networkPlacement } doReturn networkPlacement
        on { revenuePrecision } doReturn revenuePrecision
    }
    private val configuration: AppLovinSdkConfiguration = mock {
        on { countryCode } doReturn countryCode
    }
    private val appLovinSdk: AppLovinSdk = mock {
        on { configuration } doReturn configuration
    }

    private val converter = AdRevenueConverter()

    @Test
    fun convert() {
        val adRevenue = converter.convert(maxAd, appLovinSdk)
        ObjectPropertyAssertions(adRevenue)
            .checkField("adRevenue", BigDecimal.valueOf(revenue))
            .checkField("currency", Currency.getInstance("USD"))
            .checkField("adType", ModuleAdType.REWARDED)
            .checkField("adNetwork", "some_networkName")
            .checkFieldIsNull("adUnitId")
            .checkFieldIsNull("adUnitName")
            .checkField("adPlacementId", "some_networkPlacement")
            .checkField("adPlacementName", "some_placement")
            .checkField("precision", "some_revenuePrecision")
            .checkField(
                "payload",
                mapOf(
                    "countryCode" to "RU",
                    "original_ad_type" to "REWARDED",
                    "original_source" to "ad-revenue-applovin-v12",
                    AdRevenueConstants.SOURCE_KEY to "applovin"
                )
            )
            .checkField("autoCollected", false)
            .checkAll()
    }

    @Test
    fun convertIfNetPayoutIsIllegal() {
        whenever(maxAd.revenue).thenReturn(Double.NaN)
        val adRevenue: ModuleAdRevenue = converter.convert(maxAd, appLovinSdk)
        ObjectPropertyAssertions(adRevenue)
            .checkField("adRevenue", BigDecimal.valueOf(0.0))
            .checkField("currency", Currency.getInstance("USD"))
            .checkField("adType", ModuleAdType.REWARDED)
            .checkField("adNetwork", "some_networkName")
            .checkFieldIsNull("adUnitId")
            .checkFieldIsNull("adUnitName")
            .checkField("adPlacementId", "some_networkPlacement")
            .checkField("adPlacementName", "some_placement")
            .checkField("precision", "some_revenuePrecision")
            .checkField(
                "payload",
                mapOf(
                    "countryCode" to "RU",
                    "original_ad_type" to "REWARDED",
                    "original_source" to "ad-revenue-applovin-v12",
                    AdRevenueConstants.SOURCE_KEY to "applovin"
                )
            )
            .checkField("autoCollected", false)
            .checkAll()
    }

    @Test
    fun convertAndCheckAutoAdType() {
        whenever(maxAd.format).thenReturn(null)
        converter.convert(maxAd, appLovinSdk).also { adRevenue ->
            assertThat(adRevenue.adType).isNull()
            assertThat(adRevenue.payload).isEqualTo(
                mapOf(
                    "countryCode" to "RU",
                    "original_ad_type" to "null",
                    "original_source" to "ad-revenue-applovin-v12",
                    AdRevenueConstants.SOURCE_KEY to "applovin"
                )
            )
        }

        whenever(maxAd.format).thenReturn(MaxAdFormat.BANNER)
        converter.convert(maxAd, appLovinSdk).also { adRevenue ->
            assertThat(adRevenue.adType).isEqualTo(ModuleAdType.BANNER)
            assertThat(adRevenue.payload).isEqualTo(
                mapOf(
                    "countryCode" to "RU",
                    "original_ad_type" to "BANNER",
                    "original_source" to "ad-revenue-applovin-v12",
                    AdRevenueConstants.SOURCE_KEY to "applovin"
                )
            )
        }

        whenever(maxAd.format).thenReturn(MaxAdFormat.MREC)
        converter.convert(maxAd, appLovinSdk).also { adRevenue ->
            assertThat(adRevenue.adType).isEqualTo(ModuleAdType.MREC)
            assertThat(adRevenue.payload).isEqualTo(
                mapOf(
                    "countryCode" to "RU",
                    "original_ad_type" to "MREC",
                    "original_source" to "ad-revenue-applovin-v12",
                    AdRevenueConstants.SOURCE_KEY to "applovin"
                )
            )
        }

        whenever(maxAd.format).thenReturn(MaxAdFormat.LEADER)
        converter.convert(maxAd, appLovinSdk).also { adRevenue ->
            assertThat(adRevenue.adType).isEqualTo(ModuleAdType.OTHER)
            assertThat(adRevenue.payload).isEqualTo(
                mapOf(
                    "countryCode" to "RU",
                    "original_ad_type" to "LEADER",
                    "original_source" to "ad-revenue-applovin-v12",
                    AdRevenueConstants.SOURCE_KEY to "applovin"
                )
            )
        }

        whenever(maxAd.format).thenReturn(MaxAdFormat.INTERSTITIAL)
        converter.convert(maxAd, appLovinSdk).also { adRevenue ->
            assertThat(adRevenue.adType).isEqualTo(ModuleAdType.INTERSTITIAL)
            assertThat(adRevenue.payload).isEqualTo(
                mapOf(
                    "countryCode" to "RU",
                    "original_ad_type" to "INTER",
                    "original_source" to "ad-revenue-applovin-v12",
                    AdRevenueConstants.SOURCE_KEY to "applovin"
                )
            )
        }

        whenever(maxAd.format).thenReturn(MaxAdFormat.APP_OPEN)
        converter.convert(maxAd, appLovinSdk).also { adRevenue ->
            assertThat(adRevenue.adType).isEqualTo(ModuleAdType.OTHER)
            assertThat(adRevenue.payload).isEqualTo(
                mapOf(
                    "countryCode" to "RU",
                    "original_ad_type" to "APPOPEN",
                    "original_source" to "ad-revenue-applovin-v12",
                    AdRevenueConstants.SOURCE_KEY to "applovin"
                )
            )
        }

        whenever(maxAd.format).thenReturn(MaxAdFormat.REWARDED)
        converter.convert(maxAd, appLovinSdk).also { adRevenue ->
            assertThat(adRevenue.adType).isEqualTo(ModuleAdType.REWARDED)
            assertThat(adRevenue.payload).isEqualTo(
                mapOf(
                    "countryCode" to "RU",
                    "original_ad_type" to "REWARDED",
                    "original_source" to "ad-revenue-applovin-v12",
                    AdRevenueConstants.SOURCE_KEY to "applovin"
                )
            )
        }

        whenever(maxAd.format).thenReturn(MaxAdFormat.REWARDED_INTERSTITIAL)
        converter.convert(maxAd, appLovinSdk).also { adRevenue ->
            assertThat(adRevenue.adType).isEqualTo(ModuleAdType.OTHER)
            assertThat(adRevenue.payload).isEqualTo(
                mapOf(
                    "countryCode" to "RU",
                    "original_ad_type" to "REWARDED_INTER",
                    "original_source" to "ad-revenue-applovin-v12",
                    AdRevenueConstants.SOURCE_KEY to "applovin"
                )
            )
        }

        whenever(maxAd.format).thenReturn(MaxAdFormat.NATIVE)
        converter.convert(maxAd, appLovinSdk).also { adRevenue ->
            assertThat(adRevenue.adType).isEqualTo(ModuleAdType.NATIVE)
            assertThat(adRevenue.payload).isEqualTo(
                mapOf(
                    "countryCode" to "RU",
                    "original_ad_type" to "NATIVE",
                    "original_source" to "ad-revenue-applovin-v12",
                    AdRevenueConstants.SOURCE_KEY to "applovin"
                )
            )
        }
    }
}
